### PR TITLE
improve mimetype performance

### PR DESCRIPTION
**Pull request by anonrig**
An alternative to speed up `isTokenChar()` detection.

### CODE REVIEW

**Review by jasnell — APPROVED** [review]

**Review by jasnell — APPROVED** [review]

File: src/workerd/util/mimetype.c++ [hunk 1]
```diff
@@ -6,24 +6,64 @@
...
+  return result;
+}();
+
+constexpr bool isTokenChar(const uint8_t c) noexcept {
+  return token_table[c];
 }
 
-bool isQuotedStringTokenChar(const unsigned char c) {
-  return (c == '\t') || (c >= 0x20 && c <= 0x7e) || (c >= 0x80);
+constexpr std::array<uint8_t, 256> quoted_string_token_table = []() constexpr {
```
Comment (danlapid) [inline src/workerd/util/mimetype.c++:50]: Do we want this lambda to be consteval instead? just to be super duper sure it's evaluated at compiletime?

Comment (danlapid) [reply src/workerd/util/mimetype.c++:50]: Likewise on the token_table lambda

File: src/workerd/util/mimetype.c++ [hunk 2]
```diff
@@ -6,24 +6,64 @@
...
+constexpr bool isWhitespace(const char c) noexcept {
   return (c == '\r' || c == '\n' || c == '\t' || c == ' ');
 }
 
-bool isTokenChar(const unsigned char c) {
-  return c == '!' || c == '#' || c == '$' || c == '%' || c == '&' || c == '\'' ||
-         c == '*' || c == '+' || c == '\\' || c == '-' || c == '.' || c == '^' ||
-         c == '_' || c == '`' || c == '|' || c == '~' || (c >= 'A' && c <= 'Z') ||
-         (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9');
+constexpr std::array<uint8_t, 256> token_table = []() constexpr {
```
Comment (mikea) [inline src/workerd/util/mimetype.c++:19]: I didn't know you can have constexpr lambda (which seems logical in retrospect). This is cool!

File: src/workerd/util/mimetype.c++ [hunk 3]
```diff
@@ -6,24 +6,64 @@
...
+  std::array<uint8_t, 256> result{};
+
+  for (uint8_t c : { '!' ,  '#' ,  '$' ,  '%' ,  '&' ,  '\'' ,
+          '*' ,  '+' ,  '\\' ,  '-' ,  '.' ,  '^' ,
+          '_' ,  '`' ,  '|' ,  '~'}) {
+    result[c] = true;
+  }
+
+  // (c >= 'A' && c <= 'Z')
+  for (uint8_t c = 65; c <= 90; c++) {
```
Comment (mikea) [inline src/workerd/util/mimetype.c++:29]: can this be `c = 'A'` or something fpr easier readability?

File: src/workerd/util/mimetype.c++ [hunk 4]
```diff
@@ -6,24 +6,64 @@
...
 #include <workerd/util/string-buffer.h>
+#include <array>
 
 namespace workerd {
 
 namespace {
 
-bool isWhitespace(const char c) {
+constexpr bool isWhitespace(const char c) noexcept {
   return (c == '\r' || c == '\n' || c == '\t' || c == ' ');
```
Comment (mikea) [inline src/workerd/util/mimetype.c++:16]: I wonder if this will benefit from a table too?

**Review by mikea — APPROVED** [review]

File: src/workerd/util/mimetype.c++ (hunk 3 above)
Comment (anonrig) [reply src/workerd/util/mimetype.c++]: Yeah definitely. Updated the pull-request.

File: src/workerd/util/mimetype.c++ (hunk 4 above)
Comment (anonrig) [reply src/workerd/util/mimetype.c++:16]: Technically yes, but the trade-off would be the increase in the binary size. I don't have a compelling reason to add or to not add this as a table.

**Review by danlapid — APPROVED** [review]

**Pull request merged**